[Flaky test] TestReportHealthMetrics

We are getting a lot of flakes for `TestReportHealthMetrics`. It's [flaked 19 times in the last month](https://app.datadoghq.com/ci/test-branch/github.com%2FDataDog%2Fdd-trace-go/dd-trace-go/main?query=test_level%3Atest%20env%3Anone%20%40git.repository.id%3A%22github.com%2FDataDog%2Fdd-trace-go%22%20%40test.service%3Add-trace-go%20%40git.branch%3Amain&currentTab=overview&env=none&eventStack=&index=citest&start=1699210600451&end=1701802600451&paused=false)
```
Failed

=== RUN   TestReportHealthMetrics
    metrics_test.go:292: 
        	Error Trace:	D:/a/dd-trace-go/dd-trace-go/ddtrace/tracer/metrics_test.go:292
        	Error:      	Not equal: 
        	            	expected: 1
        	            	actual  : 0
        	Test:       	TestReportHealthMetrics
    metrics_test.go:293: 
        	Error Trace:	D:/a/dd-trace-go/dd-trace-go/ddtrace/tracer/metrics_test.go:293
        	Error:      	Not equal: 
        	            	expected: 1
        	            	actual  : 0
        	Test:       	TestReportHealthMetrics
--- FAIL: TestReportHealthMetrics (1.32s)
```

--- a/contrib/database/sql/sql_test.go
+++ b/contrib/database/sql/sql_test.go
@@ -38,8 +38,10 @@ func TestMain(m *testing.M) {
 		fmt.Println("--- SKIP: to enable integration test, set the INTEGRATION environment variable")
 		os.Exit(0)
 	}
-	defer sqltest.Prepare(tableName)()
-	os.Exit(m.Run())
+	cleanup := sqltest.Prepare(tableName)
+	testResult := m.Run()
+	cleanup()
+	os.Exit(testResult)
 }
 
 func TestSqlServer(t *testing.T) {

--- a/contrib/gopkg.in/jinzhu/gorm.v1/gorm_test.go
+++ b/contrib/gopkg.in/jinzhu/gorm.v1/gorm_test.go
@@ -35,8 +35,10 @@ func TestMain(m *testing.M) {
 		fmt.Println("--- SKIP: to enable integration test, set the INTEGRATION environment variable")
 		os.Exit(0)
 	}
-	defer sqltest.Prepare(tableName)()
-	os.Exit(m.Run())
+	cleanup := sqltest.Prepare(tableName)
+	testResult := m.Run()
+	cleanup()
+	os.Exit(testResult)
 }
 
 func TestMySQL(t *testing.T) {

--- a/contrib/gorm.io/gorm.v1/gorm_test.go
+++ b/contrib/gorm.io/gorm.v1/gorm_test.go
@@ -44,8 +44,10 @@ func TestMain(m *testing.M) {
 		fmt.Println("--- SKIP: to enable integration test, set the INTEGRATION environment variable")
 		os.Exit(0)
 	}
-	defer sqltest.Prepare(tableName)()
-	os.Exit(m.Run())
+	cleanup := sqltest.Prepare(tableName)
+	testResult := m.Run()
+	cleanup()
+	os.Exit(testResult)
 }
 
 func TestMySQL(t *testing.T) {

--- a/contrib/jinzhu/gorm/gorm_test.go
+++ b/contrib/jinzhu/gorm/gorm_test.go
@@ -35,8 +35,10 @@ func TestMain(m *testing.M) {
 		fmt.Println("--- SKIP: to enable integration test, set the INTEGRATION environment variable")
 		os.Exit(0)
 	}
-	defer sqltest.Prepare(tableName)()
-	os.Exit(m.Run())
+	cleanup := sqltest.Prepare(tableName)
+	testResult := m.Run()
+	cleanup()
+	os.Exit(testResult)
 }
 
 func TestSqlServer(t *testing.T) {

--- a/contrib/jmoiron/sqlx/sql_test.go
+++ b/contrib/jmoiron/sqlx/sql_test.go
@@ -29,8 +29,10 @@ func TestMain(m *testing.M) {
 		fmt.Println("--- SKIP: to enable integration test, set the INTEGRATION environment variable")
 		os.Exit(0)
 	}
-	defer sqltest.Prepare(tableName)()
-	os.Exit(m.Run())
+	cleanup := sqltest.Prepare(tableName)
+	testResult := m.Run()
+	cleanup()
+	os.Exit(testResult)
 }
 
 func TestMySQL(t *testing.T) {

--- a/ddtrace/tracer/metrics_test.go
+++ b/ddtrace/tracer/metrics_test.go
@@ -286,7 +286,7 @@ func TestReportHealthMetrics(t *testing.T) {
 
 	tracer.StartSpan("operation").Finish()
 	flush(1)
-	tg.Wait(3, 1*time.Second)
+	tg.Wait(3, 10*time.Second)
 
 	counts := tg.Counts()
 	assert.Equal(int64(1), counts["datadog.tracer.spans_started"])

--- a/ddtrace/tracer/writer.go
+++ b/ddtrace/tracer/writer.go
@@ -97,8 +97,8 @@ func (h *agentTraceWriter) flush() {
 			p.clear()
 
 			<-h.climit
-			h.wg.Done()
 			h.statsd.Timing("datadog.tracer.flush_duration", time.Since(start), nil, 1)
+			h.wg.Done()
 		}(time.Now())
 
 		var count, size int

